### ISSUE TITLE
Download API: CoreWebView2DownloadStartingEventArgs should indicate if the ResultFilePath is default or user-selected

### ISSUE THREAD

**Issue by RendijsSmukulis** [MicrosoftEdge/WebView2Feedback#1390]
The CoreWebView2DownloadStartingEventArgs event present in the pre-release versions currently does not surface information if the file download is attempting to save the file to the default download location (e.g. user clicked a link that leads to a file automatically downloading), or whether the path was chosen by the user by clicking "Save as" and choosing the download path themselves.

If we want to override the default download path by setting `ResultFilePath` we need to know whether the path was specifically chosen by the user, or is merely defaulting to the user's configured download path.

Possible ways I see to improve the API to handle this:

- Add a flag that indicates whether the path was selected by the user or not
- Add a "Default Download Path" setting to CoreWebView2.Settings
- Add a way to handle the "Save As" dialog shown event, and some "download id" to correlate the event with the actual CoreWebView2DownloadStartingEvent  event

[AB#33818732](https://microsoft.visualstudio.com/90b2a23c-cab8-4e7c-90e7-a977f32c1f5d/_workitems/edit/33818732)

**Comment by liminzhu** [comment]
Sorry jumping here kinda late. The default download location should be users' Download folder, so by comparing that and the `ResultFilePath`, you should be able to figure out if the path is default or user-selected. 

Does this solve your concern? Some of the suggestions are also interesting but if we go down those routes instead of just comparing paths, I'd love know more about what experience you're building for your end users and we can determine the right path forward.

**Comment by RendijsSmukulis** [comment]
> The default download location should be users' Download folder, so by comparing that and the ResultFilePath, you should be able to figure out if the path is default or user-selected.

I'm afraid this logic does not work if the user was presented the "Save as" dialog and they chose a path that matches the default download path. That is, we'd like to differentiate between "browser automatically downloaded to the default downloads folder" versus "user consciously chose to download to the default downloads folder".

**Comment by liminzhu** [comment]
> I'm afraid this logic does not work if the user was presented the "Save as" dialog and they chose a path that matches the default download path.

Yah that's totally fair. Can you help me understand how you'd use the information about whether it's a user selected or default  download location and the experience you're building around that? This can help inform what sort of design we want to pursue. We could do a save-as event, or a download path setting as you mentioned, but it would depend on what sort of experience we'd be supporting.

**Comment by RendijsSmukulis** [comment]
The intention is to allow our app to define a custom default downloads folder, potentially different from the system's default downloads folder.

So we're looking to implement this behavior:
- If no "save-as" dialog is presented, the file is downloaded to our custom default download path.
- If "save-as" dialog is presented, the file is downloaded to the path chosen by user, initially defaulting to the our custom default download path if user has not downloaded anything before.

For us, the "download path setting" would be a simpler approach, but an event we can react to would work just as well.

**Comment by liminzhu** [comment]
@RendijsSmukulis thanks for the context, this is very helpful! We have this on our backlog, at the mean time, can you see if this can be fulfilled by using [CDP](https://docs.microsoft.com/en-us/microsoft-edge/webview2/how-to/chromium-devtools-protocol)'s [setDownloadBehavior](https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-setDownloadBehavior)?

**Comment by RendijsSmukulis** [comment]
@liminzhu I'll give that a go, thanks! 

I suspect this will not work for devices that have devtools disabled with a policy (e.g. https://chromeenterprise.google/policies/#DeveloperToolsAvailability or https://chromeenterprise.google/policies/#DeveloperToolsDisabled ), but I'll test it.

Thank you for digging into this!

**Comment by liminzhu** [comment]
@RendijsSmukulis no prob! Devtools policy is also unlikely to affect you given WebView2 actually [isn't affected by browser policies](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/enterprise#group-policies-for-webview2).

**Comment by RendijsSmukulis** [comment]
@liminzhu This does not appear to work, sadly.

I've tried both the `DevToolsProtocolHelper` approach, and `CallDevToolsProtocolMethodAsync` approach.

```
var h = this.webview2.CoreWebView2.GetDevToolsProtocolHelper();
await h.Browser.SetDownloadBehaviorAsync("default");
await this.webview2.CoreWebView2.CallDevToolsProtocolMethodAsync("Browser.setDownloadBehavior", "{\"behavior\": \"default\"}");
```

with various combinations of values for `behavior` and `downloadPath` included/excluded. In either case I get a `System.ArgumentException: 'Value does not fall within the expected range.'`:

```
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)
   at Microsoft.Web.WebView2.Core.CoreWebView2.<CallDevToolsProtocolMethodAsync>d__150.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.Web.WebView2.Core.DevToolsProtocolExtension.Browser.<SetDownloadBehaviorAsync>d__11.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at WpfApp5.MainWindow.<<-ctor>b__0_0>d.MoveNext() in 
[..]\MainWindow.xaml.cs:line 35
```

I've verified that other commands work, such as `Browser.crash` and `Browser.getHistograms`.

This is with the latest stable (1.0.902.49) API version and pre-release (1.0.955-prerelease).
Does this API work on your end?

**Comment by liminzhu** [comment]
@RendijsSmukulis sorry about that, this does not seem to work on my end either. A couple particular domains in CDP can have issues like this. Given CDP doesn't feel like a viable alternative, we have this work item on a fairly high pri on our backlog.

**Comment by RendijsSmukulis** [comment]
@liminzhu would it be possible to get any visibility on the ETA of this?

**Comment by liminzhu** [comment]
We very recently did our planning for the next quarter. Due to an unfortunately long list of high pri asks, this is right below the line of what we think we can get to next quarter :/. I would set the expectation to be the following quarter (Q1 2022).

**Comment by champnic** [comment]
@RendijsSmukulis This is now available in 1.0.1133-prerelease+, and should be available in an upcoming SDK release.
https://docs.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2profile.defaultdownloadfolderpath?view=webview2-dotnet-1.0.1189-prerelease

Thanks!
